text/0249-v2-experiments.md

```diff
@@ -0,0 +1,127 @@
+> facing API. Experimental modules are developed in a separate repository. [link]
+
+The new aws-cdk-previews repository will have the following Readme:
+
+> This repository contains experimental code related to the AWS Cloud Development Kit (AWS CDK). CDK is an open-source
+> software development framework to define cloud infrastructure in code and provision it through AWS CloudFormation.
+> Please refer to the main repository here [link]. [TODO...]
+
+The contributor guide can be copied from the main repository, since the toolchain for building the preview repository
+will be basically the same.
+
+# Design Summary
+
+- Upon release of v2.0 to GA, all experimental modules will be automatically marked as stable (in fact, the metadata in
+  the code regarding stability can be removed, since everything will be stable). This means that the opportunity for the
+  team and for the community to influence the shape of existing experimental APIs is between now and then. Once we go
+  GA, we have to live with our design decisions.
+- Any APIs that cannot be graduated with confidence in time for the v2 release will be deprecated in 1.x, removed from
+  the v2 branch, and moved to the preview repository. This option should be minimized, since some customers may have
+  deployed production stacks that depend on these APIs.
+- Create a new repository called aws-cdk-previews that will be the new home for experimental development and developer
+  preview releases. The packages in this repository will be vended in the same way that 1.x packages are vended today,
+  in a variety of languages, with metadata that draws a line between new experiments and more mature developer preview
+  packages. This repo will take a dependency on the core CDK package just like any other independent library. This
+  avoids the complications associated with experimenting on a fork or a branch of the main repository. The repository
+  will be clearly marked as experimental.
+- When a package in the preview repository has reached maturity and is ready for GA, it will be copied into the main
+  repo and made a part of CDK 2. The module in the preview repository will then be marked as deprecated with clear
+  documentation to point users to the new stable version in the main repository. The lifecycle of experimental modules
+  in the preview repository will be exactly the same as it is currently in 1.x, with the same metadata markers in the
+  code.
+- The contributor guide for the CDK will be modified so that users know where to introduce new experimental code.
+- If it becomes absolutely necessary to add a new feature to v2 that would be a breaking change for users, the feature
+  will be hidden behind a feature flag. The code path will not be accessible without a specific opt-in by the user.
+
+# Drawbacks
+
+- In several places in the current 1.x CDK branch, stable modules refer to or contain experimental code. This will no
```

Comment:
This is listed as a drawback, but what I'd like you to do is reason through the consequences of this.

I know your focus with this doc is mostly on service libraries. Take Cognito, which is purposely only half-implemented right now. Describe to me how we would implement the other half.

Reply:
Why not create a new package in the preview repository named something like cognito_additions with the new classes? Then when we GA the additional features, it would be a simple import change. If we need to make changes on the stable repo to incorporate the new classes, we wait until they are merged to do it.

Reply:
If we do that, are any of the currently stable classes going to need to interact with types from the new library? Were we planning to do `UserPool.fromIdentityPool(...)` (or the other way around)?

If so, how would that work?

Reply:
`cognito.UserPoolClientIdentityProvider.custom("MyProviderName")]` can be used for new IDP providers developed externally. Once merged in to the stable package, we can go back and add convenience properties similar to existing ones like `public static readonly FACEBOOK = new UserPoolClientIdentityProvider('Facebook')`.

That being said, I do think we will find places where the stable package has to be modified first, in order to make it flexible enough to interact with things being developed externally, which is a drawback.